jscomp/ext/ext_ident.ml

```diff
@@ -58,18 +58,28 @@ type[@ocaml.warning "-unused-constructor"] t =
   | Global of string
   | Predef of { name: string; stamp: int }
 
-let make_js_object (i : Ident.t) =
-  (* FIXME(anmonteiro): fix for 4.12 *)
-  (* i.flags <- i.flags lor js_object_flag *)
-  (Obj.magic (Scoped { name = Ident.name i; stamp = Ident.stamp i; scope = js_object_flag }) : Ident.t)
+(* XXX(anmonteiro): This is an artifact of the 4.12 upgrade that we need to fix
+   at some point. This project (as well as the OCaml compiler, previously)
+   abuses the `Ident.t` type in order to know whether some identifier comes
+   from the JS side or not, with the objective of dealing with shadowing
+   accordingly (semantics of OCaml and JS are different, e.g. with regards to
+   recursion).
+
+   An interesting read: https://github.com/ocaml/ocaml/pull/1980 *)
+let create_unsafe_dont_use_or_bad_things_will_happen ~scope ~stamp name =
+  (Obj.magic (Scoped { name; stamp; scope }) : Ident.t)
 
-  (* Ident.create_scoped ~scope:js_flag name *)
+let make_js_object (i : Ident.t) =
+  create_unsafe_dont_use_or_bad_things_will_happen
+   ~stamp:(Ident.stamp i)
+   ~scope:(Ident.scope i lor js_object_flag)
+   (Ident.name i)
 
 (* It's a js function hard coded by js api, so when printing,
    it should preserve the name
 *)
 let create_js (name : string) : Ident.t  =
-  Ident.create_scoped ~scope:js_flag name
+  create_unsafe_dont_use_or_bad_things_will_happen ~stamp:0 ~scope:js_flag name
```

Comment:
Not sure if putting stamp 0 is a good idea, it may be because of how the ident is created, by first incrementing it then reading

https://github.com/ocaml/ocaml/blob/trunk/typing/ident.ml#L39

But then there is still the problem of two idents with the same name will have the same stamp which means they're the same ident, even if they shouldn't. I think this is a no go and we should have a stamp counter like the ident.ml file in OCaml.

Reply:
As discussed in private, this is an ident in JS, it came from the FFI, which means it is actually the same ident, and on the previous versions of the compiler it was described as a global with the flag for JS.

So this patch is right, but would love if we could improve the comment at line 78.

So `create_js` means an `js ident` while `make_js_object` is an OCaml ident that describes a JS object.